File: solutions/4014-merge-close-characters-ii/merge-close-characters-ii.java
class Solution {
    public String mergeCharacters(String s, int k) {
        StringBuilder sb = new StringBuilder();
        int[] last = new int[26];
        Arrays.fill(last, -1);

        for(char c:s.toCharArray()){
            int index = c - 'a';
            int currentPos = sb.length();

            if(last[index]!=-1&&currentPos-last[index]<=k){
                continue;
            }
            last[index] = currentPos;
            sb.append(c);
        }

        return sb.toString();
    }
}

File: solutions/4014-merge-close-characters-ii/merge-close-characters-ii.py
class Solution:
    def mergeCharacters(self, s: str, k: int) -> str:
        res, ch = [], set()
        for c in s:
            if c in ch:
                continue
            if len(ch) == k:
                ch.remove(res[-k])
            ch.add(c)
            res.append(c)
        return ''.join(res)